src/runtime_src/core/common/drv/kds_core.c

```diff
@@ -42,6 +42,32 @@ int store_kds_echo(struct kds_sched *kds, const char *buf, size_t count,
 	return count;
 }
 
+/* Each line is a CU, format:
+ * "cu_idx kernel_name:cu_name address status usage"
+ */
+ssize_t show_kds_custat_raw(struct kds_sched *kds, char *buf)
+{
+	struct kds_cu_mgmt *cu_mgmt = &kds->cu_mgmt;
+	struct xrt_cu *xcu;
+	char *cu_fmt = "%d %s:%s 0x%llx 0x%x %llu\n";
+	ssize_t sz = 0;
+	int i;
+
+	mutex_lock(&cu_mgmt->lock);
+	for (i = 0; i < cu_mgmt->num_cus; ++i) {
+		xcu = cu_mgmt->xcus[i];
+		sz += sprintf(buf+sz, cu_fmt, i, xcu->info.kname,
```

Comment:
How are you preventing buffer overruns?   
In other words, but happens if the character buffer (e.g., buf) is only 1 byte in size.  This will crash and if it happens at a privilege level, the linux kernel can be compromised.  A better way would be to:

1. Pass into this function the size of the buffer.
2. Call snprintf(NULL,....) to get the size of the print string without actually printing.
3. If the size can fit, then call sprintf()

In addition, the string character pointers should examined to make sure that they are not null before creating this status output.  

Nutshell
I assume that the sprintf() is needed for either a dmsg or some other human readable format, but it is very dangerous to use the sprintf() method and if this is not needed, we shouldn't be using it.

Reply:
Linux allocates a buffer of size (PAGE_SIZE) and passes it to this function. I don't see other driver would check if buf is null and use size as an argument in sysfs show() method.
This is documented in https://www.kernel.org/doc/Documentation/filesystems/sysfs.txt

But I would use scnprint() to prevent buffer overrun.